src/callbacks_step/time_series_dg_tree.jl

```diff
@@ -0,0 +1,185 @@
+# The coordinates are known to be located in the respective element in `element_ids`
+function calc_interpolating_polynomials!(interpolating_polynomials, coordinates,
+                                         element_ids,
+                                         mesh::TreeMesh, dg::DGSEM, cache)
+    @unpack tree = mesh
+    @unpack nodes = dg.basis
+
+    wbary = barycentric_weights(nodes)
+
+    for index in 1:length(element_ids)
+        # Construct point
+        x = SVector(ntuple(i -> coordinates[i, index], ndims(mesh)))
+
+        # Convert to unit coordinates
+        cell_id = cache.elements.cell_ids[element_ids[index]]
+        cell_coordinates_ = cell_coordinates(tree, cell_id)
+        cell_length = length_at_cell(tree, cell_id)
+        unit_coordinates = (x .- cell_coordinates_) * 2 / cell_length
+
+        # Calculate interpolating polynomial for each dimension, making use of tensor product structure
+        for d in 1:ndims(mesh)
+            interpolating_polynomials[:, d, index] .= lagrange_interpolating_polynomials(unit_coordinates[d],
+                                                                                         nodes,
+                                                                                         wbary)
+        end
+    end
+
+    return interpolating_polynomials
+end
+
+# Record the solution variables at each given point for the 1D case
+function record_state_at_points!(point_data, u, solution_variables,
+                                 n_solution_variables,
+                                 mesh::TreeMesh{1}, equations, dg::DG,
+                                 time_series_cache)
+    @unpack element_ids, interpolating_polynomials = time_series_cache
+    old_length = length(first(point_data))
+    new_length = old_length + n_solution_variables
+
+    # Loop over all points/elements that should be recorded
+    for index in 1:length(element_ids)
+        # Extract data array and element id
+        data = point_data[index]
+        element_id = element_ids[index]
+
+        # Make room for new data to be recorded
+        resize!(data, new_length)
+        data[(old_length + 1):new_length] .= zero(eltype(data))
+
+        # Loop over all nodes to compute their contribution to the interpolated values
+        for i in eachnode(dg)
+            u_node = solution_variables(get_node_vars(u, equations, dg, i,
+                                                      element_id), equations)
+
+            for v in 1:length(u_node)
+                data[old_length + v] += (u_node[v] *
+                                         interpolating_polynomials[i, 1, index])
+            end
+        end
+    end
+end
+
+# Record the solution variables at each given point for the 2D case
+function record_state_at_points!(point_data, u, solution_variables,
+                                 n_solution_variables,
+                                 mesh::TreeMesh{2},
```

Comment:
```suggestion
                                 mesh::Union{TreeMesh{2},UnstructuredMesh2D},
```
I think we can reuse this function for both of the 2D mesh types and avoid duplication in the `time_series_dg_unstructured.jl`. Although I am okay if we keep both for the sake of each file being "self-contained".

Reply:
Yes, this function is currently a duplicate. Since we decided in https://github.com/trixi-framework/Trixi.jl/pull/1873#discussion_r1526180057 to group the callback per mesh-type it is unclear where this function should be located, so I think it's best to let each mesh type have its own version.

Reply:
Okay, that makes sense to me.